### PR TITLE
[testnet1] more aggressive peering

### LINKED ISSUE #437: defunct peers are are never tried again
Scenario -

1) node running on laptop, mix of healthy and defunct peers -
```
Dec 06 09:44:19.198 DEBG monitoring_peers: 4 / 8 / 279
Dec 06 09:44:19.200 DEBG monitoring_peers: all - 279, healthy - 144, banned - 0, defunct - 135
```

2) kill wifi connection, start seeing timeouts, peers flagged as defunct - 
```
...
Dec 06 09:46:59.198 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:46:59.200 DEBG monitoring_peers: all - 279, healthy - 32, banned - 0, defunct - 247
Dec 06 09:47:09.197 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:47:09.199 DEBG monitoring_peers: all - 279, healthy - 24, banned - 0, defunct - 255
Dec 06 09:47:19.197 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:47:19.199 DEBG monitoring_peers: all - 279, healthy - 16, banned - 0, defunct - 263
Dec 06 09:47:29.197 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:47:29.199 DEBG monitoring_peers: all - 279, healthy - 8, banned - 0, defunct - 271
Dec 06 09:47:39.200 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:47:39.202 DEBG monitoring_peers: all - 279, healthy - 0, banned - 0, defunct - 279
Dec 06 09:47:49.202 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:47:49.205 DEBG monitoring_peers: all - 279, healthy - 0, banned - 0, defunct - 279
```

3) turn wifi back on, node does not try to connect to any defunct peers - 
```
Dec 06 09:50:29.198 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:50:29.201 DEBG monitoring_peers: all - 279, healthy - 0, banned - 0, defunct - 279
Dec 06 09:50:39.197 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:50:39.199 DEBG monitoring_peers: all - 279, healthy - 0, banned - 0, defunct - 279
Dec 06 09:50:49.197 DEBG monitoring_peers: 0 / 0 / 279
Dec 06 09:50:49.198 DEBG monitoring_peers: all - 279, healthy - 0, banned - 0, defunct - 279
```

4) restart grin, node attempts to connect to *some* defunct peers - 
```
Dec 06 09:54:17.336 DEBG monitoring_peers: 3 / 3 / 279
Dec 06 09:54:17.338 DEBG monitoring_peers: all - 279, healthy - 3, banned - 0, defunct - 276
Dec 06 09:54:27.335 DEBG monitoring_peers: 3 / 3 / 279
Dec 06 09:54:27.336 DEBG monitoring_peers: all - 279, healthy - 3, banned - 0, defunct - 276
Dec 06 09:54:37.336 DEBG monitoring_peers: 3 / 3 / 279
Dec 06 09:54:37.338 DEBG monitoring_peers: all - 279, healthy - 3, banned - 0, defunct - 276
```

But we are below the desired number of connected peers (8).
And for some reason we are not attempting to (re)connect to enough peers in this state.
We stay stuck with 3 connected peers, those 3 are marked healthy, but all other peers are still defunct (and will not be tried).

### PR DESCRIPTION
**Pull request by antiochp**
[WIP] looking into why so many peers get marked as defunct (do we do this too quickly?)

* attempt to peer 100 addrs at a time
* if we end up with too many then prune the connected peers down to preferred count
* rework timeout handling in `connect_and_req`
    * push timeout logic into `connect_peer` (timeout/connection error handling)
* request peer list from our conected peers when we are low on peers
    * peers that have connected to our peers may be worth trying to connect to

Resolves #437.

If this gets _really_ noisy at the network level we can revisit this but I think this is worth a shot now to see if it improves network stability.

### CODE REVIEW

**Comment by antiochp** [comment]
Some thoughts - 

* Lots of peers are actually NAT'd right now so we cannot connect directly to them
    * When these come online they connect (hopefully) to an accessible peer
* We can connect to seeds (when they are up) and a handful of other accessible nodes

If node `N1` connects successfully to seed `S1` we do not necessarily broadcast this out to the wider network (we only ask/send lists of peers on first connection unless I'm mistaken).
So I think it may be worth investigating the following - 

On successful connection, both connected peers should broadcast out their connected peers list.
This may provide some (valuable) additional info to the rest of the network about who is currently "online".

**Comment by heunglee** [comment]
This comment is at the level of learning curve. As you mentioned, currently the peer connections of grin network is shrinking quickly at this experimental stage, which is some roadblock of productive developments/tests. When we see it in the other side, the protection could be restriction sometimes. So I long for seeing antiochp's solution and evolving from there. I am looking forward to surfing on antiochp's solution.

**Comment by ignopeverell** [comment]
This sounds a bit too chatty, many people connect and leave, not to mention the possible flip/flopping internet connections at some coffee shop. On the other hand, I do agree that checking only on first connection isn't enough, particularly now.

How about re-issuing the peer-discovery message when we find out we fell below some min peer limit?

**Comment by antiochp** [comment]
Yeah - was thinking a bit more about this.
We already have a timer running that's checking for the current connected peer count.
Right now we just try and connect to some in the db.
We could easily add an additional step here that asks for peer lists from existing peers (as we do on initial connect).

File: grin/src/seed.rs [hunk 1]
```diff
@@ -286,29 +300,23 @@ fn connect_and_req(
...
 				}
-			}
-			_ => {}
+			},
+			Ok(None) => {
+				error!(LOGGER, "connect_and_req: ok but none inner (what does this mean?), {}", addr);
+			},
+			Err(e) => {
+				error!(LOGGER, "connect_and_req: err - {:?}, {}, flagging as defunct", e, addr);
+				let _ = p2p_server.update_state(addr, p2p::State::Defunct);
```
Comment (ignopeverell) [inline grin/src/seed.rs]: Logging as error will get people worried there's some real issue whereas this is bound to happen fairly often, no?

Comment (antiochp) [reply grin/src/seed.rs]: :+1: will update to `debug` level.

**Pull request merged**